# Subject: Classical Data Analysis

## Session 1 - Regression

### Individual assignment 1 Solution

Develop a regression analysis in Statmodels (with and without a constant) and SKLearn, based on the Iris sklearn dataset. This data sets consists of 3 different types of irises’ (Setosa, Versicolour, and Virginica) petal and sepal length.

See here for more information on this dataset: https://en.wikipedia.org/wiki/Iris_flower_data_set 

Use the field “sepal width (cm)” as independent variable and the field “sepal length (cm)” as dependent variable.

- Interpret and discuss the OLS Regression Results.
- Commit scripts in your GitHub account. You should export your solution code (.ipynb notebook) and push it to your repository “ClassicalDataAnalysis”.

The following are the tasks that should complete and synchronize with your repository “ClassicalDataAnalysis” until October 13. Please notice that none of these tasks is graded, however it’s important that you correctly understand and complete them in order to be sure that you won’t have problems with further assignments.

# 1 - Linear Regression in Statsmodels

## Load the iris dataset

In [1]:
import statsmodels.api as sm

In [2]:
from sklearn import datasets 

In [3]:
data = datasets.load_iris() 

In [4]:
print (data.DESCR)

Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)
    ============== ==== ==== ======= ===== ====================

    :Missing Attribute Values: None
  

In [5]:
data

{'DESCR': 'Iris Plants Database\n====================\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)\n    ============== ==== ==== ======= ===== ==

In [6]:
data.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [7]:
data.target_names

array(['setosa', 'versicolor', 'virginica'], 
      dtype='<U10')

In [8]:
import numpy as np

In [9]:
import pandas as pd

In [10]:
df = pd.DataFrame(data.data, columns=data.feature_names) 

In [11]:
df 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
5                  5.4               3.9                1.7               0.4
6                  4.6               3.4                1.4               0.3
7                  5.0               3.4                1.5               0.2
8                  4.4               2.9                1.4               0.2
9                  4.9               3.1                1.5               0.1
10                 5.4               3.7                1.5               0.2
11                 4.8               3.4                1.6               0.2
12                 4.8               3.0                1.4               0.1
13                 4.3               3.0                1.1               0.1
14                 5.8               4.0                1.2               0.2
15                 5.7               4.4                1.5               0.4
16                 5.4               3.9                1.3               0.4
17                 5.1               3.5                1.4               0.3
18                 5.7               3.8                1.7               0.3
19                 5.1               3.8                1.5               0.3
20                 5.4               3.4                1.7               0.2
21                 5.1               3.7                1.5               0.4
22                 4.6               3.6                1.0               0.2
23                 5.1               3.3                1.7               0.5
24                 4.8               3.4                1.9               0.2
25                 5.0               3.0                1.6               0.2
26                 5.0               3.4                1.6               0.4
27                 5.2               3.5                1.5               0.2
28                 5.2               3.4                1.4               0.2
29                 4.7               3.2                1.6               0.2
..                 ...               ...                ...               ...
120                6.9               3.2                5.7               2.3
121                5.6               2.8                4.9               2.0
122                7.7               2.8                6.7               2.0
123                6.3               2.7                4.9               1.8
124                6.7               3.3                5.7               2.1
125                7.2               3.2                6.0               1.8
126                6.2               2.8                4.8               1.8
127                6.1               3.0                4.9               1.8
128                6.4               2.8                5.6               2.1
129                7.2               3.0                5.8               1.6
130                7.4               2.8                6.1               1.9
131                7.9               3.8                6.4               2.0
132                6.4               2.8                5.6               2.2
133                6.3               2.8                5.1               1.5
134                6.1               2.6                5.6               1.4
135                7.7               3.0                6.1               2.3
136                6.3               3.4                5.6               2.4
137                6.4               3.1                5.5               1.8
138                6.0               3.0                4.8               1.8
139                6.9     

In [12]:
df. head() 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [13]:
df.head(5)[df.columns[0:4]] 

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

### 1.1. Regression model with Statsmodels and without a constant:

In [14]:
import statsmodels.api as sm

X = df["sepal width (cm)"]
y = df["sepal length (cm)"]

In [15]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)


In [16]:
predictions = model.predict(X) 

In [17]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:      sepal length (cm)   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     3316.
Date:                Wed, 18 Oct 2017   Prob (F-statistic):          1.04e-103
Time:                        16:50:47   Log-Likelihood:                -243.13
No. Observations:                 150   AIC:                             488.3
Df Residuals:                     149   BIC:                             491.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------------
sepal width (cm)     1.8717      0.033     57.585      0.000         1.807     1.936
==============================================================================
Omnibus:                       16.884   Durbin-Watson:                   0.429
Prob(Omnibus):                  0.000   Jarque-Bera (JB):                7.669
Skew:                          -0.336   Prob(JB):                       0.0216
Kurtosis:                       2.120   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the Table 

The coefficient of 1.8717 means that as the sepal width (cm variable increases by 1, the predicted value of sepal length (cm) increases by 1.8717. We can see here that this model has a much high R-squared value 0.957, meaning that this model explains 95.7% of the variance in our dependent variable. 

### 1.2. Regression model with Statsmodels and with a constant:

In [18]:
import statsmodels.api as sm 
X = df["sepal width (cm)"]
y = df["sepal length (cm)"]

In [19]:
X = sm.add_constant(X) 

In [20]:
model = sm.OLS(y, X).fit() 

In [21]:
predictions = model.predict(X)

In [22]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:      sepal length (cm)   R-squared:                       0.012
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     1.792
Date:                Wed, 18 Oct 2017   Prob (F-statistic):              0.183
Time:                        16:50:47   Log-Likelihood:                -183.14
No. Observations:                 150   AIC:                             370.3
Df Residuals:                     148   BIC:                             376.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------------
const                6.4812      0.481     13.466      0.000         5.530     7.432
sepal width (cm)    -0.2089      0.156     -1.339      0.183        -0.517     0.099
==============================================================================
Omnibus:                        4.455   Durbin-Watson:                   0.941
Prob(Omnibus):                  0.108   Jarque-Bera (JB):                4.252
Skew:                           0.356   Prob(JB):                        0.119
Kurtosis:                       2.585   Cond. No.                         24.3
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the Table 
With the constant term the coefficients are different. Without a constant we are forcing our model to go through the origin, but now we have a y-intercept at 6.4812. We also changed the slope of the sepal width (cm) predictor from 1.8717 to -0.2089.  A few other important values are modification of the R-squared, from 0.957 to 0.012.

# 2 - Linear Regression in SKLearn 

In [48]:
from sklearn import linear_model

In [49]:
from sklearn import datasets 

In [50]:
data = datasets.load_iris() 

In [51]:
df = pd.DataFrame(data.data, columns=data.feature_names)

In [52]:
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
5                  5.4               3.9                1.7               0.4
6                  4.6               3.4                1.4               0.3
7                  5.0               3.4                1.5               0.2
8                  4.4               2.9                1.4               0.2
9                  4.9               3.1                1.5               0.1
10                 5.4               3.7                1.5               0.2
11                 4.8               3.4                1.6               0.2
12                 4.8               3.0                1.4               0.1
13                 4.3               3.0                1.1               0.1
14                 5.8               4.0                1.2               0.2
15                 5.7               4.4                1.5               0.4
16                 5.4               3.9                1.3               0.4
17                 5.1               3.5                1.4               0.3
18                 5.7               3.8                1.7               0.3
19                 5.1               3.8                1.5               0.3
20                 5.4               3.4                1.7               0.2
21                 5.1               3.7                1.5               0.4
22                 4.6               3.6                1.0               0.2
23                 5.1               3.3                1.7               0.5
24                 4.8               3.4                1.9               0.2
25                 5.0               3.0                1.6               0.2
26                 5.0               3.4                1.6               0.4
27                 5.2               3.5                1.5               0.2
28                 5.2               3.4                1.4               0.2
29                 4.7               3.2                1.6               0.2
..                 ...               ...                ...               ...
120                6.9               3.2                5.7               2.3
121                5.6               2.8                4.9               2.0
122                7.7               2.8                6.7               2.0
123                6.3               2.7                4.9               1.8
124                6.7               3.3                5.7               2.1
125                7.2               3.2                6.0               1.8
126                6.2               2.8                4.8               1.8
127                6.1               3.0                4.9               1.8
128                6.4               2.8                5.6               2.1
129                7.2               3.0                5.8               1.6
130                7.4               2.8                6.1               1.9
131                7.9               3.8                6.4               2.0
132                6.4               2.8                5.6               2.2
133                6.3               2.8                5.1               1.5
134                6.1               2.6                5.6               1.4
135                7.7               3.0                6.1               2.3
136                6.3               3.4                5.6               2.4
137                6.4               3.1                5.5               1.8
138                6.0               3.0                4.8               1.8
139                6.9     

In [53]:
df2 = pd.DataFrame(df, columns=["sepal width (cm)"])

In [54]:
df2

sepal width (cm)
0                 3.5
1                 3.0
2                 3.2
3                 3.1
4                 3.6
5                 3.9
6                 3.4
7                 3.4
8                 2.9
9                 3.1
10                3.7
11                3.4
12                3.0
13                3.0
14                4.0
15                4.4
16                3.9
17                3.5
18                3.8
19                3.8
20                3.4
21                3.7
22                3.6
23                3.3
24                3.4
25                3.0
26                3.4
27                3.5
28                3.4
29                3.2
..                ...
120               3.2
121               2.8
122               2.8
123               2.7
124               3.3
125               3.2
126               2.8
127               3.0
128               2.8
129               3.0
130               2.8
131               3.8
132               2.8
133               2.8
134               2.6
135               3.0
136               3.4
137               3.1
138               3.0
139               3.1
140               3.1
141               3.1
142               2.7
143               3.2
144               3.3
145               3.0
146               2.5
147               3.0
148               3.4
149               3.0

[150 rows x 1 columns]

In [55]:
target = pd.DataFrame(df, columns=["sepal length (cm)"])

In [56]:
target

sepal length (cm)
0                  5.1
1                  4.9
2                  4.7
3                  4.6
4                  5.0
5                  5.4
6                  4.6
7                  5.0
8                  4.4
9                  4.9
10                 5.4
11                 4.8
12                 4.8
13                 4.3
14                 5.8
15                 5.7
16                 5.4
17                 5.1
18                 5.7
19                 5.1
20                 5.4
21                 5.1
22                 4.6
23                 5.1
24                 4.8
25                 5.0
26                 5.0
27                 5.2
28                 5.2
29                 4.7
..                 ...
120                6.9
121                5.6
122                7.7
123                6.3
124                6.7
125                7.2
126                6.2
127                6.1
128                6.4
129                7.2
130                7.4
131                7.9
132                6.4
133                6.3
134                6.1
135                7.7
136                6.3
137                6.4
138                6.0
139                6.9
140                6.7
141                6.9
142                5.8
143                6.8
144                6.7
145                6.7
146                6.3
147                6.5
148                6.2
149                5.9

[150 rows x 1 columns]

In [57]:
X = df2
y = target

In [58]:
lm = linear_model.LinearRegression()
model = lm.fit(X,y)

In [59]:
type(predictions)

numpy.ndarray

In [60]:
predictions = lm.predict(X)
print(predictions[0:5,])

[[ 5.75017718]
 [ 5.85461233]
 [ 5.81283827]
 [ 5.8337253 ]
 [ 5.72929015]]


In [61]:
print(predictions)

[[ 5.75017718]
 [ 5.85461233]
 [ 5.81283827]
 [ 5.8337253 ]
 [ 5.72929015]
 [ 5.66662906]
 [ 5.77106421]
 [ 5.77106421]
 [ 5.87549936]
 [ 5.8337253 ]
 [ 5.70840312]
 [ 5.77106421]
 [ 5.85461233]
 [ 5.85461233]
 [ 5.64574204]
 [ 5.56219392]
 [ 5.66662906]
 [ 5.75017718]
 [ 5.68751609]
 [ 5.68751609]
 [ 5.77106421]
 [ 5.70840312]
 [ 5.72929015]
 [ 5.79195124]
 [ 5.77106421]
 [ 5.85461233]
 [ 5.77106421]
 [ 5.75017718]
 [ 5.77106421]
 [ 5.81283827]
 [ 5.8337253 ]
 [ 5.77106421]
 [ 5.62485501]
 [ 5.60396798]
 [ 5.8337253 ]
 [ 5.81283827]
 [ 5.75017718]
 [ 5.8337253 ]
 [ 5.85461233]
 [ 5.77106421]
 [ 5.75017718]
 [ 6.00082154]
 [ 5.81283827]
 [ 5.75017718]
 [ 5.68751609]
 [ 5.85461233]
 [ 5.68751609]
 [ 5.81283827]
 [ 5.70840312]
 [ 5.79195124]
 [ 5.81283827]
 [ 5.81283827]
 [ 5.8337253 ]
 [ 6.00082154]
 [ 5.89638639]
 [ 5.89638639]
 [ 5.79195124]
 [ 5.97993451]
 [ 5.87549936]
 [ 5.91727342]
 [ 6.06348262]
 [ 5.85461233]
 [ 6.02170856]
 [ 5.87549936]
 [ 5.87549936]
 [ 5.8337253 ]
 [ 5.85461

In [62]:
lm.score(X,y) 

0.011961632834767699

In [63]:
lm.coef_ ## will give this output:

array([[-0.20887029]])

In [64]:
lm.intercept_ ## that will give this output:

array([ 6.48122321])